管理者権限の付与機能と退職者のパラメータを追加

rhistoba opened the pull request.

rhistoba committed:
ユーザー登録と編集に管理者権限を付与できるようにした
また管理者が１人だけのときその管理者は削除できないようにした

rhistoba committed:
ja.ymlに関連する項目を追加

rhistoba committed:
Userにretireカラムを追加

rhistoba committed:
ユーザー情報で退職者を設定できるようにした
ユーザー情報ページで管理者と退職者のパラメータを見えるようにした

rhistoba committed:
退職者の編集に関するテストを追加

rhistoba committed:
退職社のログインを無効化、退職者の日報をトップページで見えないようにした

rhistoba committed:
管のみ退職者のユーザーページにアクセスできるよう修正

rhistoba committed:
users_controllerのindexアクションを修正

rhistoba committed:
ユーザー作成時のフォームに退職者の項目を出さないよう変更

ore-public commented on db/migrate/20170529024013_add_retire_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddRetireToUsers < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :retire, :boolean, default: false
```
```ruby
null: false
```
も追加しよう。

http://qiita.com/jnchito/items/a342b64cd998e5c4ef3d

ore-public commented on app/controllers/sessions_controller.rb:
```diff
@@ -5,8 +5,13 @@ def new
-      redirect_to root_path
+      if user.retire?
+        flash.now[:danger] = t('flash.session.create.retire')
+        render 'new'
```
ここだけのチェックだと、ログイン中の状態で
別のユーザーに退職扱いされたときに、
ログアウトするまでは操作出来てしまう気がする。

ore-public commented on app/controllers/users_controller.rb:
```diff
@@ -4,10 +4,15 @@ class UsersController < ApplicationController
+    if current_user.admin?
+      @users = User.paginate(page: params[:page])
+    else
+      @users = User.where(retire: false).paginate(page: params[:page])
```
```ruby
User.where(retire: false)
```
という部分が他の箇所と重複してる。

こういう時は、scopeを使って共通化するのがいいです。ちょっと調べてみて。

ore-public commented on app/controllers/users_controller.rb:
```diff
@@ -38,16 +43,23 @@ def update
-    User.find(params[:id]).destroy
-    flash[:success] = t('flash.user.destroy.success')
+    user = User.find(params[:id])
+    # user.admin==false なら && 以降は実行されない
```
このコメントは、Rubyの文法についての解説になってしまっているので、ここにあるのはふさわしくない。
削除でいいかな。

ore-public commented on app/controllers/users_controller.rb:
```diff
@@ -38,16 +43,23 @@ def update
-    flash[:success] = t('flash.user.destroy.success')
+    user = User.find(params[:id])
+    # user.admin==false なら && 以降は実行されない
+    if user.admin? && User.where(admin: true).count <= 1
```
こういう複合条件の時は、どういう条件かを表す名前を付けたメソッドを定義するといい。

ore-public reviewed (changes requested):
@rhistoba 以上です

rhistoba committed:
booleanのカラムに対してnull: falseを追加

rhistoba committed:
退職者ユーザーによる操作を無効化

rhistoba committed:
scopeを使って在職中のユーザーを取り出す処理を共通化

rhistoba committed:
不要なコメントを削除

rhistoba committed:
users_controllerのdestroyアクションの複合条件のメソッド化と条件の追加

rhistoba committed:
ユーザーの編集と削除に関する分岐を追加

管理者が1人だけのとき、そのユーザーは削除されず、adminを解除できない
管理者かつ在職中のユーザーが1人だけの時、そのユーザーは削除されず、退職者とならない

rhistoba committed:
ユーザーの編集に関する分岐を追加

管理者かつ在職中のユーザーが1人しかいないとき、そのユーザーは退職者扱いになれず管理者にもなれない

rhistoba committed:
チェックボックスの挙動を修正

ore-public commented on app/controllers/comments_controller.rb:
```diff
@@ -1,7 +1,8 @@
   before_action :confirm_editable_user, only: :destroy
-  before_action :set_report, only: [:create, :destroy]
+  before_action :set_report
+  before_action :confirm_retire
```
confirm_loginと、confirm_retireは大半のコントローラで必要になる処理で
不要なcontorollerの方が少ない。
基本的には認証が必要な機能がベースのシステムなので、この二つはApplicationControllerで設定してやって
認証が不要なcontrollerでのみ、skip_before_actionを書くほうが安全。
デフォルトの挙動を安全側に倒すのがセキュリティ上重要です。

ore-public commented on app/controllers/users_controller.rb:
```diff
@@ -60,4 +69,41 @@ def set_user
+
+  def admin_to_not_admin?
+    @user.admin? && (user_params[:admin]=='false')
+  end
```
less_admin_user?以降のメソッドは、役割としてモデルに有るべき。
Userの状態に関する判定処理だから。

後で、画面上でコードをいじりながらレクチャーします。

rhistoba committed:
loginとretireの確認のbefore_actionをapplication_controllerに移動

rhistoba committed:
adminとretireの分岐で条件の被りを修正

rhistoba committed:
less_working_admin?をusers_controllerからuser.rbに移動

rhistoba committed:
Userモデルに編集と削除のバリデーションを追加

コントローラー側で行なっていた処理をモデル側に移動

ore-public commented on app/models/user.rb:
```diff
@@ -21,9 +23,25 @@ def self.digest(string)
+    if (self.admin_was && !self.retire_was) &&
+        (!self.admin? || self.retire?) &&
+        less_working_admin?
+      self.errors[:base] << "在職中の管理者は1人以上必要です"
```
Modelの中でも
```ruby
I18n.t('validates.user.user_over_create')
```
みたいに、localeを呼べるのでそれを使おう。

rhistoba committed:
userモデルのバリデーションのエラーメッセージをlocaleを使うよう修正
